### PR TITLE
A variety of improvements around tool parameter modeling (second try)

### PR DESCRIPTION
**Pull request by jmchilton**
A redo of https://github.com/galaxyproject/galaxy/pull/18952 on top of the new tool execution testing stuff and with a few more fixes. 

Old PR Description:

This contains tests, fixes, and refactorings of the tool parameter models aimed at enabling two applications/APIs - these applications are a tool landing API and the tool request API. Those APIs and corresponding runtimes are not included here but can be found in the structured tool state branch ([tool request API](https://github.com/galaxyproject/galaxy/pull/17393/commits/e717c335437614522e4db9e1cb58ba63a9b22f7a) and [tool landing API](https://github.com/galaxyproject/galaxy/pull/17393/commits/c3fa15b5d100efa6df5f7496070236bbe913cf18)).

The bulk of the enhancements around enabling tool landing are found in a single commit titled "[models for tool landing request state...](https://github.com/galaxyproject/galaxy/commit/bcf9fab1c75fb0a720de3ecbb35a490a3af7f7bf)". This introduces two new tool state representation types and wide variety of parameter specification tests that describe how the API will work and how models will be stored. The API will consume tool state with the representation type "landing_request" and will store the representation "landing_request_internal" in the database. The internal version is the decoded version of that tool state. The idea behind the landing API is that essentially everything is optional - external entities could for instance host a collection of FASTA files and provide links to tools to align against these for instance - all the parameters but the database would be deferred to the users discretion. The parameters that are provided though will be validated.

The remainder of the commits are a collection of things that make more confident were not doing damage with the tool request API. Part of that effort is having a fully expanded representation of the tool state that describes something pretty close to what we feed the wrappers - defaults expanded out, no absent conditionals, etc... That effort resulting in me writing lots of API tests to describe how the current API works so we don't regress it, implementing validation of static parameter validators so we can mimic some of the runtime behavior of the tools and write stronger upfront validation of the requests, etc...

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

File: lib/galaxy/tool_util/parser/parameter_validators.py [hunk 1]
```diff
@@ -0,0 +1,725 @@
...
+    models = []
+    for validator_el in validator_els:
+        models.append(parse_xml_validator(validator_el))
+    return models
+
+
+def static_validators(validator_models: List[AnyValidatorModel]) -> List[AnyValidatorModel]:
+    static_validators = []
+    for validator_model in validator_models:
+        print(validator_model._static)
```
Comment (bgruening) [inline lib/galaxy/tool_util/parser/parameter_validators.py:472]:
```suggestion
```

File: lib/galaxy/tools/parameters/basic.py [hunk 2]
```diff
@@ -404,9 +394,16 @@ class TextToolParameter(SimpleTextToolParameter):
     def __init__(self, tool, input_source):
         input_source = ensure_input_source(input_source)
         super().__init__(tool, input_source)
+        self.profile = tool.profile if tool else None
         self.datalist = []
         for title, value, _ in input_source.parse_static_options():
             self.datalist.append({"label": title, "value": value})
+
+        # why does Integer and Float subclass this :_(
```
Comment (bgruening) [inline lib/galaxy/tools/parameters/basic.py:402]: why does it? :)

File: lib/galaxy/tools/parameters/validation.py [hunk 3]
```diff
@@ -391,35 +299,19 @@ class MetadataInFileColumnValidator(Validator):
...
+        metadata_name: str,
+        metadata_column: int,
+        message: str,
+        line_startswith: Optional[str] = None,
+        split: str = "\t",
+        negate: bool = False,
     ):
         super().__init__(message, negate)
+        assert filename
+        assert os.path.exists(filename), f"File {filename} specified by the 'filename' attribute not found"
```
Comment (bgruening) [inline lib/galaxy/tools/parameters/validation.py:314]: Is that really needed here as assert?

File: lib/galaxy/tools/stock.py [hunk 4]
```diff
@@ -26,7 +26,7 @@ def stock_tool_sources():
 
 
 def _walk_directory_for_tools(path):
-    if path.is_file() and path.name.endswith(".xml"):
+    if path.is_file() and path.name.endswith(".xml") and "macros" not in path.name and "_conf.xml" not in path.name:
```
Comment (bgruening) [inline lib/galaxy/tools/stock.py]: can you add a comment what you are trying here?

File: lib/galaxy_test/base/populators.py [hunk 5]
```diff
@@ -3603,6 +3617,34 @@ def execute(self) -> "DescribeToolExecution":
...
+        if self._input_format == "legacy":
+            self._inputs = inputs
+        return self
+
+    def nested(self, inputs: Dict[str, Any]) -> Self:
+        if self._input_format == "21.01":
+            self._inputs = inputs
+        return self
+
+    # aliases for self to create silly little English sentense... inputs.when.flat().when.legacy()
```
Comment (bgruening) [inline lib/galaxy_test/base/populators.py:3648]: :-)

File: test/functional/tools/parameters/gx_drill_down_code.xml [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        <import>macros.xml</import>
+    </macros>
+    <code file="gx_drill_down_code.py"/>
+    <command><![CDATA[
+echo "parameter: $parameter" > '$output'
+    ]]></command>
+    <inputs>
+        <param name="ref_parameter" type="data" ext="tabular" >
+        </param>
+g        <param name="parameter" type="drill_down" dynamic_options="collate_table(ref_parameter.dataset.get_file_name())">
```
Comment (bgruening) [inline test/functional/tools/parameters/gx_drill_down_code.xml:12]:
```suggestion
        <param name="parameter" type="drill_down" dynamic_options="collate_table(ref_parameter.dataset.get_file_name())">
```
How could this pass tests?

File: lib/galaxy/tools/stock.py (hunk 4 above)
Comment (jmchilton) [reply lib/galaxy/tools/stock.py]: Added a comment.

File: lib/galaxy/tools/parameters/validation.py (hunk 3 above)
Comment (jmchilton) [reply lib/galaxy/tools/parameters/validation.py:314]: It was present in the existing validation code. I guess the idea is to not load the tool if that file is not available.

File: lib/galaxy/tools/parameters/basic.py (hunk 2 above)
Comment (jmchilton) [reply lib/galaxy/tools/parameters/basic.py:402]: I don't know but the code sharing in this file is very problematic all over.

**Review by bgruening — APPROVED** [review]

**Review by bgruening — APPROVED** [review]
Thanks @jmchilton

**Pull request merged**